src/gausskernel/storage/access/k2/k2_table_ops.cpp

```diff
@@ -265,70 +255,20 @@ static bool IsSystemCatalogChange(Relation rel)
 	return IsSystemRelation(rel) && !IsBootstrapProcessingMode();
 }
 
-/*
- * Utility method to execute a prepared write statement.
- * Will handle the case if the write changes the system catalogs meaning
- * we need to increment the catalog versions accordingly.
- */
-static void K2PgExecWriteStmt(K2PgStatement k2pg_stmt, Relation rel, int *rows_affected_count)
-{
-	bool is_syscatalog_change = IsSystemCatalogChange(rel);
-	bool modifies_row = false;
-	HandleK2PgStatus(PgGate_DmlModifiesRow(k2pg_stmt, &modifies_row));
-
-	/*
-	 * If this write may invalidate catalog cache tuples (i.e. UPDATE or DELETE),
-	 * or this write may insert into a cached list, we must increment the
-	 * cache version so other sessions can invalidate their caches.
-	 * NOTE: If this relation caches lists, an INSERT could effectively be
-	 * UPDATINGing the list object.
-	 */
-	bool is_syscatalog_version_change = is_syscatalog_change && modifies_row;
-	//		&& (modifies_row || RelationHasCachedLists(rel));
-
-	/* Let the master know if this should increment the catalog version. */
-	if (is_syscatalog_version_change)
-	{
-		HandleK2PgStatus(PgGate_SetIsSysCatalogVersionChange(k2pg_stmt));
```

Comment:
It is added to each Exec call now with a bool flag that is set like "bool increment_catalog = IsSystemCatalogChange(rel);". The cached version is incremented after each exec call. This way we can make it transactional with the change. I removed the condition on modifies_row, which made it so that Inserts did not increment the catalog version which seemed wrong to me, but I can change that back if you think Inserts don't need to change the version.

Reply:
The insert logic was covered by the RelationHasCachedLists(rel) call, which I disabled it temporarily. I will put back the logic later. Please keep modifies_row for now. Thanks

Reply:
Looks like modifies_row was just set to true if the statement was an UPDATE or DELETE, and always false for INSERT. I will add a TODO mentioning RelationHasCachedLists(rel) on the insert path for a reminder.